Design/SANS/NewISISDesign/Design_Document.md

```diff
@@ -0,0 +1,350 @@
+|------|------------|
+|ADS   | Analysis Data Service|
+| PI   | Python Interface |
+
+
+## Motivation
+
+The current SANS reduction system makes use of the so-called `ReductionSingleton`,
+which essentially stores the configurational state either directly, in `ReductionStep`s
+or in an `ISISInstrument` object. The `ReductionStep`s are deeply coupled to the
+`ReductionSingleton`, which makes it hard to unit test them
+(and probably explains why such tests do not exist).
+In addition there is a strong coupling of this `ReductionSingleton` to the
+GUI logic, which essentially renders the ISIS SANS reduction a monolithic block.
+
+A more modern approach is to use Mantid's `WorkflowAlgorithm`s. A sequence of
+algorithms perform the reduction, where each algorithm gets the information
+required for the reduction at the time it executes. The `WorkflowAlgorithm` is
+used as a unit of clearly defined work and **not** as a container for
+state information. This mechanism avoids the deep coupling as each algorithm is
+agnostic about its calling environment and the information flow is into the algorithms
+(except for resulting output workspaces).
+This creates an ideal environment for unit testing and improved documentation.
+
+
+## Reducer 2.0
+
+As per requirement [R.2.4](User_Requirements.md#R.2.4) the reduction of a single
+run should be the limit case of the reduction of a batch file. Currently the
+implementations are slightly different. Unifying the two approaches is beneficial to future maintenance.
+
+### SANS State
+
+The current ISIS SANS system suffers from a state which is distributed and stored
+in many different places. The state for a the SANS reduction system should be
+ localized and complete, i.e. it defines the entire reduction.
+
+A design document for a `SANSState` approach is available
+[here](SANSState/SANSState.md).
+
+The result of this document is that the `SANSState` object contains a
+complete definition of a single reduction (provided the state is valid, which should
+  be checked by validators adequate for the facility/instrument).
+
+### Batch reduction
+
+Starting point of a reduction is a collection of `SANSState` objects which define one
+ reduction per `SANSState`. This batch reduction is handled by a work-flow
+ algorithm `SANSBatchReduction`. It should be able to operate the individual
+ reductions in parallel (does this mean, we need to implement it in `C++`?),
+ which is in line with requirement [R.3.2](#User_Requirements#R.3.2).
+
+The `SANSBatchReduction` performs two tasks:
+
+1. Loads files into the ADS if they are not present yet. These workspaces remain
+in mint condition, i.e. we don't want to move them in order to be able to reuse
+ them. The calibration workspace is being loaded into the ADS as well.
+2. Passes the loaded workspaces and the `SANSState` to a work-flow algorithm,
+which performs a single reduction.
+
+The above ADS feature might be considered unclean, since a work-flow algorithm
```

Comment:
The Reflectometry workflow does this, and I consider it to be a mistake at some level. I would strongly suggest that if you want to cache something in the ADS you have an abstract way of looking at something in a `SansCache` which sits over the ADS rather than coding against the ADS directly. Direct ADS access violates the current coding model for Algorithms, and will break the future (forthcoming pythonic API)